experimental/metal/pipeline_layout.m

```diff
@@ -0,0 +1,180 @@
+
+typedef struct iree_hal_metal_descriptor_set_layout_t {
+  // Abstract resource used for injecting reference counting and vtable; must be at offset 0.
+  iree_hal_resource_t resource;
+
+  iree_allocator_t host_allocator;
+
+  iree_host_size_t binding_count;
+  iree_hal_descriptor_set_layout_binding_t bindings[];
+} iree_hal_metal_descriptor_set_layout_t;
+
+static const iree_hal_descriptor_set_layout_vtable_t iree_hal_metal_descriptor_set_layout_vtable;
+
+static iree_hal_metal_descriptor_set_layout_t* iree_hal_metal_descriptor_set_layout_cast(
+    iree_hal_descriptor_set_layout_t* base_value) {
+  IREE_HAL_ASSERT_TYPE(base_value, &iree_hal_metal_descriptor_set_layout_vtable);
+  return (iree_hal_metal_descriptor_set_layout_t*)base_value;
+}
+
+iree_status_t iree_hal_metal_descriptor_set_layout_create(
+    iree_allocator_t host_allocator, iree_hal_descriptor_set_layout_flags_t flags,
+    iree_host_size_t binding_count, const iree_hal_descriptor_set_layout_binding_t* bindings,
+    iree_hal_descriptor_set_layout_t** out_descriptor_set_layout) {
+  IREE_ASSERT_ARGUMENT(!binding_count || bindings);
+  IREE_ASSERT_ARGUMENT(out_descriptor_set_layout);
+  *out_descriptor_set_layout = NULL;
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+  iree_hal_metal_descriptor_set_layout_t* descriptor_set_layout = NULL;
+  iree_host_size_t bindings_size = binding_count * sizeof(descriptor_set_layout->bindings[0]);
+  iree_status_t status =
+      iree_allocator_malloc(host_allocator, sizeof(*descriptor_set_layout) + bindings_size,
+                            (void**)&descriptor_set_layout);
+  if (iree_status_is_ok(status)) {
+    iree_hal_resource_initialize(&iree_hal_metal_descriptor_set_layout_vtable,
+                                 &descriptor_set_layout->resource);
+    descriptor_set_layout->host_allocator = host_allocator;
+    descriptor_set_layout->binding_count = binding_count;
+    memcpy(descriptor_set_layout->bindings, bindings, bindings_size);
+    *out_descriptor_set_layout = (iree_hal_descriptor_set_layout_t*)descriptor_set_layout;
+  }
+  IREE_TRACE_ZONE_END(z0);
+  return status;
+}
+
+static void iree_hal_metal_descriptor_set_layout_destroy(
+    iree_hal_descriptor_set_layout_t* base_descriptor_set_layout) {
+  iree_hal_metal_descriptor_set_layout_t* descriptor_set_layout =
+      iree_hal_metal_descriptor_set_layout_cast(base_descriptor_set_layout);
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+  iree_allocator_free(descriptor_set_layout->host_allocator, descriptor_set_layout);
+
+  IREE_TRACE_ZONE_END(z0);
+}
+
+iree_hal_descriptor_set_layout_binding_t* iree_hal_metal_descriptor_set_layout_binding(
+    iree_hal_descriptor_set_layout_t* base_descriptor_set_layout, uint32_t binding) {
+  iree_hal_metal_descriptor_set_layout_t* descriptor_set_layout =
+      iree_hal_metal_descriptor_set_layout_cast(base_descriptor_set_layout);
+  for (unsigned i = 0, e = descriptor_set_layout->binding_count; i < e; ++i) {
```

Comment:
proper types, and no need for caching e as this is a basic read-only loop of a struct variable and it doesn't help